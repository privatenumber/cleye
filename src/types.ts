import type {
	TypeFlag,
	TypeFlagOptions,
	Flags as BaseFlags,
} from 'type-flag';
import { Command } from './command';
import type { Renderers } from './render-help/renderers';

export declare const parsedType: unique symbol;

export type Flags = BaseFlags<{

	/**
	Description to be used in help output

	@example
	```
	description: 'Unit of output (metric, imperial)',
	```
	*/
	description?: string;

	/**
	Placeholder label to be used in help output

	@example Required value
	```
	placeholder: '<locale>'
	```
	*/
	placeholder?: string;
}>;

export type CallbackFunction<Parsed> = (parsed: {
	// This exposes the content of "TypeFlag<T>" in type hints
	[Key in keyof Parsed]: Parsed[Key];
}) => void;

type HasVersion<Options extends { flags?: Flags }> = (
	Options extends { version: string }
		? Options['flags'] & { version: BooleanConstructor }
		: Options['flags']
);

type HasHelp<Options extends { flags?: Flags }> = (
	Options extends { help: false }
		? Options['flags']
		: Options['flags'] & { help: BooleanConstructor }
);

export type HasHelpOrVersion<Options extends { flags?: Flags }> = (
	HasVersion<Options> & HasHelp<Options>
);

export type HelpDocumentNode<Types extends PropertyKey = keyof Renderers> = {
	id?: string;
	type: Types;
	data: any;
};

export type HelpOptions = {

	/**
	Version of the script displayed in `--help` output. Use to avoid enabling `--version` flag.
	*/
	version?: string;

	/**
	Description of the script or command to display in `--help` output.
	*/
	description?: string;

	/**
	Usage code examples to display in `--help` output.
	*/
	usage?: false | string | string[];

	/**
	Example code snippets to display in `--help` output.
	*/
	examples?: string | string[];

	/**
	Function to customize the help document before it is logged.
	*/
	render?: (
		nodes: HelpDocumentNode<keyof Renderers>[],
		renderers: Renderers,
	) => string;
};

export type CliOptions<
	Commands = Command[],
	Parameters extends string[] = string[],
> = {
	/**
	Name of the script displayed in `--help` output.
	*/
	name?: string;

	/**
	Version of the script displayed in `--version` and `--help` outputs.
	*/
	version?: string;

	/**
	Parameters accepted by the script. Parameters must be in the following formats:

	- Required parameter: `<parameter name>`
	- Optional parameter: `[parameter name]`
	- Required spread parameter: `<parameter name...>`
	- Optional spread parameter: `[parameter name...]`
	*/
	parameters?: Parameters;

	/**
	Commands to register to the script.
	*/
	commands?: Commands;

	/**
	Flags accepted by the script
	*/
	flags?: Flags;

	/**
	Options to configure the help documentation. Pass in `false` to disable handling `--help, -h`.
	*/
	help?: false | HelpOptions;

	/**
	 * Which argv elements to ignore from parsing
	 */
	ignoreArgv?: TypeFlagOptions['ignore'];
};

export type CliOptionsInternal<
	Commands = Command[],
> = CliOptions<Commands> & {
	parent?: CliOptions;
};

type AlphabetLowercase = 'a' | 'b' | 'c' | 'd' | 'e' | 'f' | 'g' | 'h' | 'i' | 'j' | 'k' | 'l' | 'm' | 'n' | 'o' | 'p' | 'q' | 'r' | 's' | 't' | 'u' | 'v' | 'w' | 'x' | 'y' | 'z';
type Numeric = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';
type AlphaNumeric = AlphabetLowercase | Uppercase<AlphabetLowercase> | Numeric;

type CamelCase<Word extends string> = (
  Word extends `${infer FirstCharacter}${infer Rest}`
    ? (
      FirstCharacter extends AlphaNumeric
        ? `${FirstCharacter}${CamelCase<Rest>}`
        : Capitalize<CamelCase<Rest>>
    )
    : Word
);

type StripBrackets<Parameter extends string> = (
	Parameter extends `<${infer ParameterName}>` | `[${infer ParameterName}]`
		? (
			ParameterName extends `${infer SpreadName}...`
				? SpreadName
				: ParameterName
		)
		: never
);

type ParameterType<Parameter extends string> = (
	Parameter extends `<${infer _ParameterName}...>` | `[${infer _ParameterName}...]`
		? string[]
		: Parameter extends `<${infer _ParameterName}>`
			? string
			: Parameter extends `[${infer _ParameterName}]`
				? string | undefined
				: never
);

type WithCommand<
	Options extends { flags?: Flags },
	Command extends string | undefined = undefined,
> = {
	command: Command;
} & Options;

type TypeFlagWrapper<
	Options extends { flags?: Flags },
	Parameters extends string[],
> = TypeFlag<HasHelpOrVersion<Options>> & {
	_: {
		[
			Parameter in Parameters[number]
				as CamelCase<StripBrackets<Parameter>>
		]: ParameterType<Parameter>;
	};
	showHelp: (options?: HelpOptions) => void;
	showVersion: () => void;
}

export type ParseArgv<
	Options extends { flags?: Flags },
	Parameters extends string[],
	Command extends string | undefined = '',
> = (
	Command extends ''
		? TypeFlagWrapper<Options, Parameters>
		: WithCommand<TypeFlagWrapper<Options, Parameters>, Command>
);
