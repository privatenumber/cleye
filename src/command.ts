import type { TypeFlagOptions } from 'type-flag';
import type {
	CallbackFunction,
	Flags,
	HelpOptions,
	ParseArgv,
	parsedType,
} from './types';
import { isScriptNamePattern } from './utils/is-script-name';

export type CommandOptions<Parameters = string[]> = {
	/**
	Name of the command used to invoke it. Also displayed in `--help` output.
	*/
	name: string;

	/**
	Aliases for the command used to invoke it. Also displayed in `--help` output.
	*/
	alias?: string | string[];

	/**
	Parameters accepted by the command. Parameters must be in the following formats:

	- Required parameter: `<parameter name>`
	- Optional parameter: `[parameter name]`
	- Required spread parameter: `<parameter name...>`
	- Optional spread parameter: `[parameter name...]`
	*/
	parameters?: Parameters;

	/**
	Flags accepted by the command
	*/
	flags?: Flags;

	/**
	Options to configure the help documentation. Pass in `false` to disable handling `--help, -h`.
	*/
	help?: false | HelpOptions;

	/**
	 * Which argv elements to ignore from parsing
	 */
	ignoreArgv?: TypeFlagOptions['ignore'];
};

export function command<
	Options extends CommandOptions<[...Parameters]>,
	Parameters extends string[],
>(
	options: Readonly<Options> & CommandOptions<[...Parameters]>,
	callback?: CallbackFunction<ParseArgv<Options, Parameters>>,
): Command<Options, ParseArgv<Options, Parameters, Options['name']>> {
	if (!options) {
		throw new Error('Command options are required');
	}

	const { name } = options;
	if (options.name === undefined) {
		throw new Error('Command name is required');
	}

	if (!isScriptNamePattern.test(name)) {
		throw new Error(`Invalid command name ${JSON.stringify(name)}. Command names must be one word.`);
	}

	// @ts-expect-error Missing parsedType, a type only property
	return {
		options,
		callback,
	};
}

export type Command<
	Options extends CommandOptions = CommandOptions,
	ParsedType = any,
> = {
	readonly options: Options;
	readonly callback?: CallbackFunction<any>;
	[parsedType]: ParsedType;
};
