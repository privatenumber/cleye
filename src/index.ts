export { cli } from './cli';
export { command, type Command } from './command';
export type { Renderers } from './render-help';
export type { Flags } from './types';
