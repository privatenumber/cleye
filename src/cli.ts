import typeFlag from 'type-flag';
import type {
	CallbackFunction,
	HasHelpOrVersion,
	CliOptions,
	CliOptionsInternal,
	ParseArgv,
	parsedType,
	HelpOptions,
	HelpDocumentNode,
} from './types';
import { Command } from './command';
import { generateHelp, Renderers } from './render-help';
import { camelCase } from './utils/convert-case';
import { isScriptNamePattern } from './utils/is-script-name';

const { stringify } = JSON;

const specialCharactersPattern = /[|\\{}()[\]^$+*?.]/;

type ParsedParameter = {
	name: string;
	required: boolean;
	spread: boolean;
};

function parseParameters(parameters: string[]) {
	const parsedParameters: ParsedParameter[] = [];

	let hasOptional: string | undefined;
	let hasSpread: string | undefined;

	for (const parameter of parameters) {
		if (hasSpread) {
			throw new Error(`Invalid parameter: Spread parameter ${stringify(hasSpread)} must be last`);
		}

		const firstCharacter = parameter[0];
		const lastCharacter = parameter[parameter.length - 1];

		let required: boolean | undefined;
		if (firstCharacter === '<' && lastCharacter === '>') {
			required = true;

			if (hasOptional) {
				throw new Error(`Invalid parameter: Required parameter ${stringify(parameter)} cannot come after optional parameter ${stringify(hasOptional)}`);
			}
		}

		if (firstCharacter === '[' && lastCharacter === ']') {
			required = false;
			hasOptional = parameter;
		}

		if (required === undefined) {
			throw new Error(`Invalid parameter: ${stringify(parameter)}. Must be wrapped in <> (required parameter) or [] (optional parameter)`);
		}

		let name = parameter.slice(1, -1);

		const spread = name.slice(-3) === '...';

		if (spread) {
			hasSpread = parameter;
			name = name.slice(0, -3);
		}

		const invalidCharacter = name.match(specialCharactersPattern);
		if (invalidCharacter) {
			throw new Error(`Invalid parameter: ${stringify(parameter)}. Invalid character found ${stringify(invalidCharacter[0])}`);
		}

		parsedParameters.push({
			name,
			required,
			spread,
		});
	}

	return parsedParameters;
}

function mapParametersToArguments(
	mapping: Record<string, string | string[]>,
	parameters: ParsedParameter[],
	cliArguments: string[],
	showHelp: () => void,
) {
	for (let i = 0; i < parameters.length; i += 1) {
		const { name, required, spread } = parameters[i];
		const camelCaseName = camelCase(name);
		if (camelCaseName in mapping) {
			throw new Error(`Invalid parameter: ${stringify(name)} is used more than once.`);
		}

		const value = spread ? cliArguments.slice(i) : cliArguments[i];

		if (spread) {
			i = parameters.length;
		}

		if (
			required
			&& (!value || (spread && value.length === 0))
		) {
			console.error(`Error: Missing required parameter ${stringify(name)}\n`);
			showHelp();
			return process.exit(1);
		}

		mapping[camelCaseName] = value;
	}
}

function helpEnabled(help: false | undefined | HelpOptions): help is (HelpOptions | undefined) {
	return (
		// Default
		help === undefined

		// Configured
		|| help !== false
	);
}

function cliBase<
	CommandName extends string | undefined,
	Options extends CliOptionsInternal,
	Parameters extends string[],
>(
	command: CommandName,
	options: Options,
	callback: CallbackFunction<ParseArgv<Options, Parameters>> | undefined,
	argv: string[],
) {
	const flags = { ...options.flags };
	const isVersionEnabled = options.version;

	// Expected to work even if flag is overwritten; add tests
	if (isVersionEnabled) {
		flags.version = {
			type: Boolean,
			description: 'Show version',
		};
	}

	const { help } = options;
	const isHelpEnabled = helpEnabled(help);

	// Expected to work even if overwritten; add tests
	if (isHelpEnabled && !('help' in flags)) {
		flags.help = {
			type: Boolean,
			alias: 'h',
			description: 'Show help',
		};
	}

	const parsed = typeFlag(flags as HasHelpOrVersion<Options>, argv);

	const showVersion = () => {
		console.log(options.version);
	};

	if (
		isVersionEnabled

		// Can be overridden to different type
		&& parsed.flags.version === true
	) {
		showVersion();
		return process.exit(0);
	}

	const helpRenderers = new Renderers();
	const renderHelpFunction = (
		(isHelpEnabled && help?.render)
			? help.render
			: (nodes: HelpDocumentNode | HelpDocumentNode[]) => helpRenderers.render(nodes)
	);

	const showHelp = (helpOptions?: HelpOptions) => {
		const nodes = generateHelp({
			...options,
			...(helpOptions ? { help: helpOptions } : {}),
			flags,
		});

		console.log(renderHelpFunction(nodes, helpRenderers));
	};

	if (
		isHelpEnabled

		// Can be overridden to different type
		&& parsed.flags.help === true
	) {
		showHelp();
		return process.exit(0);
	}

	if (options.parameters) {
		let { parameters } = options;
		let cliArguments = parsed._ as string[];
		const hasEof = parameters.indexOf('--');
		const eofParameters = parameters.slice(hasEof + 1);
		const mapping: Record<string, string | string[]> = Object.create(null);

		if (hasEof > -1 && eofParameters.length > 0) {
			parameters = parameters.slice(0, hasEof);

			const eofArguments = parsed._['--'];
			cliArguments = cliArguments.slice(0, -eofArguments.length || undefined);

			mapParametersToArguments(
				mapping,
				parseParameters(parameters),
				cliArguments,
				showHelp,
			);
			mapParametersToArguments(
				mapping,
				parseParameters(eofParameters),
				eofArguments,
				showHelp,
			);
		} else {
			mapParametersToArguments(
				mapping,
				parseParameters(parameters),
				cliArguments,
				showHelp,
			);
		}

		Object.assign(
			parsed._,
			mapping,
		);
	}

	const parsedWithApi = {
		...parsed,
		showVersion,
		showHelp,
	};

	if (typeof callback === 'function') {
		callback(parsedWithApi as any);
	}

	// Already flattened
	return {
		command,
		...parsedWithApi,
	};
}

function getCommand<Commands extends Command[]>(
	potentialCommand: string,
	commands: [...Commands],
) {
	const commandMap = new Map<string, Command>();
	for (const command of commands) {
		const names = [command.options.name];

		const { alias } = command.options;
		if (alias) {
			if (Array.isArray(alias)) {
				names.push(...alias);
			} else {
				names.push(alias);
			}
		}

		for (const name of names) {
			if (commandMap.has(name)) {
				throw new Error(`Duplicate command name found: ${stringify(name)}`);
			}

			commandMap.set(name, command);
		}
	}

	return commandMap.get(potentialCommand);
}

function cli<
	Options extends CliOptions<undefined, [...Parameters]>,
	Parameters extends string[],
>(
	options: Options & CliOptions<undefined, [...Parameters]>,
	callback?: CallbackFunction<ParseArgv<Options, Parameters>>,
	argv?: string[],
): {
	[
		Key in keyof ParseArgv<
			Options,
			Parameters,
			undefined
		>
	]: ParseArgv<
		Options,
		Parameters,
		undefined
	>[Key];
};

function cli<
	Options extends CliOptions<[...Commands], [...Parameters]>,
	Commands extends Command[],
	Parameters extends string[],
>(
	options: Options & CliOptions<[...Commands], [...Parameters]>,
	callback?: CallbackFunction<ParseArgv<Options, Parameters>>,
	argv?: string[],
):(
	{
		[
			Key in keyof ParseArgv<
				Options,
				Parameters,
				undefined
			>
		]: ParseArgv<
			Options,
			Parameters,
			undefined
		>[Key];
	}
	| {
		[KeyA in keyof Commands]: (
			Commands[KeyA] extends Command
				? (
					{
						[
							KeyB in keyof Commands[KeyA][typeof parsedType]
						]: Commands[KeyA][typeof parsedType][KeyB];
					}
				) : never
		);
	}[number]
);

function cli<
	Options extends CliOptions<[...Commands], [...Parameters]>,
	Commands extends Command[],
	Parameters extends string[],
>(
	options: Options | (Options & CliOptions<[...Commands], [...Parameters]>),
	callback?: CallbackFunction<ParseArgv<Options, Parameters>>,
	argv = process.argv.slice(2),
) {
	// Because if not configured, it's probably being misused or overlooked
	if (!options) {
		throw new Error('Options is required');
	}

	if ('name' in options && (!options.name || !isScriptNamePattern.test(options.name))) {
		throw new Error(`Invalid script name: ${stringify(options.name)}`);
	}

	const potentialCommand = argv[0];

	if (
		options.commands
		&& isScriptNamePattern.test(potentialCommand)
	) {
		const command = getCommand(potentialCommand, options.commands);

		if (command) {
			return cliBase(
				command.options.name,
				{
					...command.options,
					parent: options,
				},
				command.callback,
				argv.slice(1),
			);
		}
	}

	return cliBase(undefined, options, callback, argv);
}

export { cli };
