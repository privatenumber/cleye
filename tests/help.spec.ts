import { underline } from 'colorette';
import { cli, command } from '../src';

let mockProcessExit: jest.SpyInstance;
let mockConsoleLog: jest.SpyInstance;

beforeAll(() => {
	process.stdout.columns = 90;
	mockProcessExit = jest.spyOn(process, 'exit').mockImplementation();
});

beforeEach(() => {
	mockConsoleLog = jest.spyOn(console, 'log').mockImplementation();

	mockProcessExit.mockClear();
});

afterEach(() => {
	mockConsoleLog.mockRestore();
});

afterAll(() => {
	mockProcessExit.mockRestore();
});

describe('show help', () => {
	test('empty cli', () => {
		cli(
			{},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('name', () => {
		cli(
			{
				name: 'npm',
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('empty parameters', () => {
		cli(
			{
				parameters: [],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('parameters with no name', () => {
		cli(
			{
				parameters: ['<arg-a>', '[arg-b]'],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('parameters with name', () => {
		cli(
			{
				name: 'my-cli',
				parameters: ['<arg-a>', '[arg-b]'],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('empty commands', () => {
		cli(
			{
				commands: [],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('commands', () => {
		const testCommand = command({
			name: 'test',
		});

		cli(
			{
				name: 'my-cli',
				commands: [
					testCommand,
				],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('commands with description', () => {
		const testCommand = command({
			name: 'test',
			help: {
				description: 'test command',
			},
		});

		cli(
			{
				name: 'my-cli',
				commands: [
					testCommand,
				],
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('undefined flags', () => {
		cli(
			{
				flags: undefined,
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('empty flags', () => {
		cli(
			{
				flags: {},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('flags', () => {
		cli(
			{
				flags: {
					flag: Boolean,
					flagA: String,
					flagB: {
						type: Number,
					},
					flagC: {
						type: RegExp,
						default: /hello/,
					},
				},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('help disabled', () => {
		cli(
			{
				help: false,
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).not.toHaveBeenCalled();
	});

	test('help disabled but shown', () => {
		const argv = cli(
			{
				name: 'my-cli',
				help: false,
			},
			undefined,
			['--help'],
		);

		argv.showHelp({
			version: '1.2.3',
		});

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('empty help.examples', () => {
		cli(
			{
				help: {
					examples: [],
				},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('help.version with --help', () => {
		cli(
			{
				help: {
					version: '1.0.0',
				},
			},
			undefined,
			['--help'],
		);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('help.version with --version', () => {
		const parsed = cli(
			{
				help: {
					version: '1.0.0',
				},
			},
			undefined,
			['--version'],
		);

		expect(mockProcessExit).not.toHaveBeenCalled();

		expect(parsed.unknownFlags).toStrictEqual({
			version: [true],
		});
	});

	test('help.usage string', () => {
		cli(
			{
				help: {
					usage: 'usage string',
				},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('help.usage array', () => {
		cli(
			{
				help: {
					usage: [
						'usage string a',
						'usage string b',
						'usage string c',
					],
				},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('help.description', () => {
		cli(
			{
				help: {
					description: 'test description',
				},
			},
			undefined,
			['--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('command help', () => {
		const testCommand = command({
			name: 'test',
			parameters: ['<arg a>', '<arg b>'],
			help: {
				description: 'test command',
			},
		});

		cli(
			{
				name: 'my-cli',
				commands: [
					testCommand,
				],
			},
			undefined,
			['test', '--help'],
		);

		expect(mockProcessExit).toHaveBeenCalledWith(0);

		const { calls } = mockConsoleLog.mock;
		expect(calls[0][0]).toMatchSnapshot();
	});

	test('command help disabled', () => {
		const testCommand = command({
			name: 'test',
			help: false,
		});

		cli(
			{
				name: 'my-cli',
				commands: [
					testCommand,
				],
			},
			undefined,
			['test', '--help'],
		);

		expect(mockProcessExit).not.toHaveBeenCalled();
	});
});

test('show version', () => {
	cli(
		{
			version: '1.0.0',
		},
		undefined,
		['--version'],
	);

	expect(mockProcessExit).toHaveBeenCalledWith(0);

	const { calls } = mockConsoleLog.mock;
	expect(calls[0][0]).toMatchSnapshot();
});

test('smoke test', () => {
	cli({
		name: 'my-cli',

		version: '1.1.1',

		commands: [
			command({
				name: 'my-command',
				help: {
					description: 'my command description',
				},
			}),
		],

		parameters: ['<urls...>'],

		flags: {
			outputDir: {
				type: String,
				alias: 'o',
				description: 'Tweet screenshot output directory',
				placeholder: '<path>',
			},
			width: {
				type: Number,
				alias: 'w',
				description: 'Width of tweet',
				default: 550,
				placeholder: '<width>',
			},
			showTweet: {
				type: Boolean,
				alias: 't',
				description: 'Show tweet thread',
			},
			darkMode: {
				type: Boolean,
				alias: 'd',
				description: 'Show tweet in dark mode',
			},
			locale: {
				type: String,
				description: 'Locale',
				default: 'en',
				placeholder: '<locale>',
			},
		},

		help: {
			examples: [
				'# Snapshot a tweet',
				'snap-tweet https://twitter.com/jack/status/20',
				'',
				'# Snapshot a tweet with Japanese locale',
				'snap-tweet https://twitter.com/TwitterJP/status/578707432 --locale ja',
				'',
				'# Snapshot a tweet with dark mode and 900px width',
				'snap-tweet https://twitter.com/Interior/status/463440424141459456 --width 900 --dark-mode',
			],
		},
	}, undefined, ['--help']);

	expect(mockProcessExit).toHaveBeenCalledWith(0);

	const { calls } = mockConsoleLog.mock;
	expect(calls[0][0]).toMatchSnapshot();
});

test('help customization', () => {
	const simpleFlags = {
		bundle: {
			type: Boolean,
			description: 'Bundle all dependencies into the output files',
		},
		define: {
			type: String,
			description: 'Substitute K with V while parsing',
		},
	};

	const advancedFlags = {
		allowOverwrite: {
			type: Boolean,
			description: 'Allow output files to overwrite input files',
		},
		assetNames: {
			type: Boolean,
			description: 'Path template to use for "file" loader files (default "[name]-[hash]")',
		},
	};

	cli({
		name: 'esbuild',

		version: '1.0.0',

		parameters: ['[entry points]'],

		flags: {
			...simpleFlags,
			...advancedFlags,

			// Overwrite to remove alias
			help: Boolean,
		},

		help: {
			examples: [
				'# Produces dist/entry_point.js and dist/entry_point.js.map',
				'esbuild --bundle entry_point.js --outdir=dist --minify --sourcemap',
			],

			render(nodes, renderers) {
				// Remove name
				nodes.splice(0, 1);

				// Replace "flags" with "options" in Usage
				nodes[0].data.body = nodes[0].data.body.replace('flags', 'options');

				// Add Documentation & Repository
				nodes.splice(1, 0, {
					type: 'section',
					data: {
						title: 'Documentation:',
						body: underline('https://esbuild.github.io/'),
					},
				});

				// Split Flags into "Simple options" and "Advanced options"
				const flags = nodes[2].data.body.data;
				nodes.splice(2, 1, {
					type: 'section',
					data: {
						title: 'Simple options:',
						body: {
							type: 'table',
							data: {
								...flags,
								tableData: flags.tableData.filter(
									([flagName]: [{ data: { name: string }}]) => flagName.data.name in simpleFlags,
								),
							},
						},
						indentBody: 0,
					},
				}, {
					type: 'section',
					data: {
						title: 'Advanced options:',
						body: {
							type: 'table',
							data: {
								...flags,
								tableData: flags.tableData.filter(
									([flagName]: [{ data: { name: string }}]) => flagName.data.name in advancedFlags,
								),
							},
						},
						indentBody: 0,
					},
				});

				// Update renderer so flags that accept a value shows `=...`
				renderers.flagOperator = () => '=';
				renderers.flagParameter = flagType => (flagType === Boolean ? '' : '...');

				const extendedRenderer = Object.assign(renderers, {
					someRenderer(value: number) {
						return `Received value: ${value}`;
					},
				});

				return extendedRenderer.render([
					...nodes,
					{
						type: 'someRenderer',
						data: 123,
					},
				]);
			},
		},
	}, undefined, ['--help']);

	expect(mockProcessExit).toHaveBeenCalledWith(0);

	const { calls } = mockConsoleLog.mock;
	expect(calls[0][0]).toMatchSnapshot();
});
