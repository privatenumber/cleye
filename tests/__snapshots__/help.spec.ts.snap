// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`help customization 1`] = `
"USAGE:
  esbuild [options...] [entry points]

DOCUMENTATION:
  [4mhttps://esbuild.github.io/[24m

SIMPLE OPTIONS:
  --bundle            Bundle all dependencies into the output files                       
  --define=...        Substitute K with V while parsing                                   

ADVANCED OPTIONS:
  --allow-overwrite        Allow output files to overwrite input files                    
  --asset-names            Path template to use for \\"file\\" loader files (default          
                           \\"[name]-[hash]\\")                                               

EXAMPLES:
  # Produces dist/entry_point.js and dist/entry_point.js.map
  esbuild --bundle entry_point.js --outdir=dist --minify --sourcemap

Received value: 123"
`;

exports[`show help command help 1`] = `
"my-cli test

test command

USAGE:
  my-cli test [flags...] <arg a> <arg b>

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help commands 1`] = `
"my-cli

USAGE:
  my-cli [flags...]
  my-cli <command>

COMMANDS:
  test                                                                                    

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help commands with description 1`] = `
"my-cli

USAGE:
  my-cli [flags...]
  my-cli <command>

COMMANDS:
  test        test command                                                                

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help empty cli 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help empty commands 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help empty flags 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help empty help.examples 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help empty parameters 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help flags 1`] = `
"FLAGS:
      --flag                                                                              
      --flag-a <string>                                                                   
      --flag-b <number>                                                                   
      --flag-c <value>          (default: {})                                             
  -h, --help                   Show help                                                  
"
`;

exports[`show help help disabled but shown 1`] = `
"my-cli v1.2.3
"
`;

exports[`show help help.description 1`] = `
"test description

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help help.usage array 1`] = `
"USAGE:
  usage string a
  usage string b
  usage string c

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help help.usage string 1`] = `
"USAGE:
  usage string

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help help.version with --help 1`] = `
"v1.0.0

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help name 1`] = `
"npm

USAGE:
  npm [flags...]

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help parameters with name 1`] = `
"my-cli

USAGE:
  my-cli [flags...] <arg-a> [arg-b]

FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help parameters with no name 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show help undefined flags 1`] = `
"FLAGS:
  -h, --help        Show help                                                             
"
`;

exports[`show version 1`] = `"1.0.0"`;

exports[`smoke test 1`] = `
"my-cli v1.1.1

USAGE:
  my-cli [flags...] <urls...>
  my-cli <command>

COMMANDS:
  my-command        my command description                                                

FLAGS:
  -d, --dark-mode                Show tweet in dark mode                                  
  -h, --help                     Show help                                                
      --locale <locale>          Locale (default: \\"en\\")                                   
  -o, --output-dir <path>        Tweet screenshot output directory                        
  -t, --show-tweet               Show tweet thread                                        
      --version                  Show version                                             
  -w, --width <width>            Width of tweet (default: 550)                            

EXAMPLES:
  # Snapshot a tweet
  snap-tweet https://twitter.com/jack/status/20
  
  # Snapshot a tweet with Japanese locale
  snap-tweet https://twitter.com/TwitterJP/status/578707432 --locale ja
  
  # Snapshot a tweet with dark mode and 900px width
  snap-tweet https://twitter.com/Interior/status/463440424141459456 --width 900 --dark-mode
"
`;
